File: wp-content/prerf/analyze-har.py
#!/usr/bin/env python3
"""
HAR File Analyzer
Analyzes HAR files for duplicate requests, errors, and resource optimization
"""

import json
import os
from urllib.parse import urlparse, parse_qs
from collections import defaultdict, Counter

class HarAnalyzer:
    def __init__(self, har_file):
        self.file_name = os.path.basename(har_file).replace('.har', '')
        with open(har_file, 'r', encoding='utf-8') as f:
            self.har_data = json.load(f)
        
        if 'log' not in self.har_data or 'entries' not in self.har_data['log']:
            raise ValueError("Invalid HAR file format")
    
    def analyze(self):
        entries = self.har_data['log']['entries']
        
        analysis = {
            'fileName': self.file_name,
            'totalRequests': len(entries),
            'duplicateRequests': [],
            'errorResponses': [],
            'preloadDuplicates': [],
            'resourcesByType': defaultdict(lambda: {'count': 0, 'totalSize': 0}),
            'wastedBytes': 0,
            'fontDuplicates': [],
            'queryStringDuplicates': []
        }
        
        # Group requests by URL and analyze
        url_groups = defaultdict(list)
        base_url_groups = defaultdict(list)
        preload_urls = set()
        font_urls = Counter()
        
        for entry in entries:
            request = entry['request']
            response = entry['response']
            url = request['url']
            method = request['method']
            
            # Parse URL
            parsed_url = urlparse(url)
            base_url = f"{parsed_url.scheme}://{parsed_url.netloc}{parsed_url.path}"
            query_string = parsed_url.query or ''
            
            # Check for preload/preconnect
            is_preload = False
            for header in request.get('headers', []):
                if header.get('name', '').lower() == 'purpose' and header.get('value') == 'prefetch':
                    is_preload = True
                    break
            
            # Determine resource type
            resource_type = self.get_resource_type(url, response)
            
            # Calculate size
            body_size = response.get('bodySize', 0)
            headers_size = response.get('headersSize', 0)
            total_size = body_size + headers_size
            
            # Track URLs
            url_groups[url].append({
                'method': method,
                'status': response['status'],
                'size': total_size,
                'mimeType': response.get('content', {}).get('mimeType', 'unknown'),
                'resourceType': resource_type,
                'isPreload': is_preload
            })
            
            # Track base URLs
            base_url_groups[base_url].append({
                'fullUrl': url,
                'queryString': query_string,
                'status': response['status'],
                'size': total_size
            })
            
            # Track preloads
            if is_preload:
                preload_urls.add(url)
            
            # Track fonts
            if resource_type == 'font':
                font_urls[url] += 1
            
            # Track errors
            if response['status'] >= 400:
                analysis['errorResponses'].append({
                    'url': url,
                    'status': response['status'],
                    'statusText': response.get('statusText', ''),
                    'resourceType': resource_type
                })
            
            # Count resources by type
            analysis['resourcesByType'][resource_type]['count'] += 1
            analysis['resourcesByType'][resource_type]['totalSize'] += total_size
        
        # Find duplicate requests
        for url, requests in url_groups.items():
            if len(requests) > 1:
                total_size = sum(req['size'] for req in requests)
                wasted_size = total_size - requests[0]['size']
                
                analysis['duplicateRequests'].append({
                    'url': url,
                    'count': len(requests),
                    'resourceType': requests[0]['resourceType'],
                    'totalSize': total_size,
                    'wastedBytes': wasted_size,
                    'requests': requests
                })
                
                analysis['wastedBytes'] += wasted_size
                
                # Check preload duplicates
                if url in preload_urls:
                    analysis['preloadDuplicates'].append({
                        'url': url,
                        'timesLoaded': len(requests),
                        'resourceType': requests[0]['resourceType']
                    })
        
        # Find query string duplicates
        for base_url, requests in base_url_groups.items():
            if len(requests) > 1:
                unique_query_strings = set(req['queryString'] for req in requests)
                
                if len(unique_query_strings) > 1:
                    analysis['queryStringDuplicates'].append({
                        'baseUrl': base_url,
                        'variants': [
                            {
                                'url': req['fullUrl'],
                                'queryString': req['queryString'],
                                'size': req['size']
                            }
                            for req in requests
                        ]
                    })
        
        # Find font duplicates
        for url, count in font_urls.items():
            if count > 1:
                analysis['fontDuplicates'].append({
                    'url': url,
                    'timesLoaded': count
                })
        
        return analysis
    
    def get_resource_type(self, url, response):
        mime_type = response.get('content', {}).get('mimeType', '')
        extension = os.path.splitext(urlparse(url).path)[1].lower()
        
        # Check by MIME type
        if 'javascript' in mime_type or 'ecmascript' in mime_type:
            return 'js'
        elif 'css' in mime_type:
            return 'css'
        elif 'image/' in mime_type:
            return 'image'
        elif 'font/' in mime_type or 'application/font' in mime_type:
            return 'font'
        elif 'text/html' in mime_type:
            return 'html'
        elif 'application/json' in mime_type:
            return 'json'
        
        # Check by extension
        ext_map = {
            '.js': 'js', '.mjs': 'js',
            '.css': 'css',
            '.jpg': 'image', '.jpeg': 'image', '.png': 'image', 
            '.gif': 'image', '.svg': 'image', '.webp': 'image', '.ico': 'image',
            '.woff': 'font', '.woff2': 'font', '.ttf': 'font', 
            '.otf': 'font', '.eot': 'font',
            '.html': 'html', '.htm': 'html',
            '.json': 'json',
            '.xml': 'xml'
        }
        
        return ext_map.get(extension, 'other')
    
    def format_report(self, analysis):
        report = f"\n=== HAR Analysis Report: {analysis['fileName']} ===\n\n"
        
        # Summary
        report += "SUMMARY:\n"
        report += f"- Total Requests: {analysis['totalRequests']}\n"
        report += f"- Duplicate Requests: {len(analysis['duplicateRequests'])}\n"
        report += f"- Error Responses (4xx/5xx): {len(analysis['errorResponses'])}\n"
        report += f"- Total Wasted Bytes: {analysis['wastedBytes']:,} bytes ({self.format_bytes(analysis['wastedBytes'])})\n\n"
        
        # Resources by Type
        report += "RESOURCES BY TYPE:\n"
        for res_type, data in sorted(analysis['resourcesByType'].items()):
            report += f"- {res_type.upper()}: {data['count']} requests, {self.format_bytes(data['totalSize'])}\n"
        report += "\n"
        
        # Duplicate Requests
        if analysis['duplicateRequests']:
            report += "DUPLICATE REQUESTS:\n"
            for dup in sorted(analysis['duplicateRequests'], key=lambda x: x['wastedBytes'], reverse=True):
                report += f"- {self.truncate_url(dup['url'])}\n"
                report += f"  Type: {dup['resourceType']}, Loaded: {dup['count']} times, Wasted: {self.format_bytes(dup['wastedBytes'])}\n"
            report += "\n"
        
        # Query String Duplicates
        if analysis['queryStringDuplicates']:
            report += "QUERY STRING DUPLICATES (same file, different parameters):\n"
            for dup in analysis['queryStringDuplicates']:
                report += f"- Base URL: {self.truncate_url(dup['baseUrl'])}\n"
                for variant in dup['variants']:
                    report += f"  * {variant['queryString'] or '[no query string]'} ({self.format_bytes(variant['size'])})\n"
            report += "\n"
        
        # Preload Duplicates
        if analysis['preloadDuplicates']:
            report += "PRELOAD DUPLICATES:\n"
            for dup in analysis['preloadDuplicates']:
                report += f"- {self.truncate_url(dup['url'])} (Type: {dup['resourceType']}, Loaded: {dup['timesLoaded']} times)\n"
            report += "\n"
        
        # Font Duplicates
        if analysis['fontDuplicates']:
            report += "FONT DUPLICATES:\n"
            for font in analysis['fontDuplicates']:
                report += f"- {self.truncate_url(font['url'])} (Loaded: {font['timesLoaded']} times)\n"
            report += "\n"
        
        # Error Responses
        if analysis['errorResponses']:
            report += "ERROR RESPONSES:\n"
            for error in analysis['errorResponses']:
                report += f"- {error['status']} {error['statusText']}: {self.truncate_url(error['url'])} (Type: {error['resourceType']})\n"
            report += "\n"
        
        return report
    
    @staticmethod
    def format_bytes(bytes_val):
        if bytes_val < 1024:
            return f"{bytes_val} B"
        elif bytes_val < 1048576:
            return f"{bytes_val / 1024:.2f} KB"
        else:
            return f"{bytes_val / 1048576:.2f} MB"
    
    @staticmethod
    def truncate_url(url, max_length=80):
        if len(url) <= max_length:
            return url
        
        parsed = urlparse(url)
        path = parsed.path
        filename = os.path.basename(path)
        
        if len(filename) < 40:
            return '...' + url[-(max_length - 3):]
        else:
            return '...' + filename[-(max_length - 3):]


def main():
    har_files = [
        'trang-chu.har',
        'trang-san-pham.har',
        'trang-bai-viet.har',
        'trang-contact.har'
    ]
    
    input_dir = os.path.join(os.path.dirname(__file__), 'inputs')
    all_reports = ''
    
    for har_file in har_files:
        file_path = os.path.join(input_dir, har_file)
        
        if not os.path.exists(file_path):
            print(f"Warning: HAR file not found: {file_path}")
            continue
        
        try:
            analyzer = HarAnalyzer(file_path)
            analysis = analyzer.analyze()
            report = analyzer.format_report(analysis)
            
            print(report)
            all_reports += report
            
        except Exception as e:
            print(f"Error analyzing {har_file}: {str(e)}")
    
    # Save combined report
    report_path = os.path.join(os.path.dirname(__file__), 'har-analysis-report.txt')
    with open(report_path, 'w', encoding='utf-8') as f:
        f.write(all_reports)
    
    print(f"\nFull report saved to: {report_path}")


if __name__ == '__main__':
    main()

File: wp-content/prerf/analyze-har.php
<?php
/**
 * HAR File Analyzer
 * Analyzes HAR files for duplicate requests, errors, and resource optimization
 */

class HarAnalyzer {
    private $harData;
    private $fileName;
    
    public function __construct($harFile) {
        $this->fileName = basename($harFile, '.har');
        $content = file_get_contents($harFile);
        $this->harData = json_decode($content, true);
        
        if (!$this->harData || !isset($this->harData['log']['entries'])) {
            throw new Exception("Invalid HAR file format");
        }
    }
    
    public function analyze() {
        $entries = $this->harData['log']['entries'];
        $analysis = [
            'fileName' => $this->fileName,
            'totalRequests' => count($entries),
            'duplicateRequests' => [],
            'errorResponses' => [],
            'preloadDuplicates' => [],
            'resourcesByType' => [],
            'wastedBytes' => 0,
            'fontDuplicates' => [],
            'queryStringDuplicates' => []
        ];
        
        // Group requests by URL and analyze
        $urlGroups = [];
        $baseUrlGroups = []; // Group by base URL without query string
        $preloadUrls = [];
        $fontUrls = [];
        
        foreach ($entries as $entry) {
            $request = $entry['request'];
            $response = $entry['response'];
            $url = $request['url'];
            $method = $request['method'];
            
            // Parse URL
            $parsedUrl = parse_url($url);
            $baseUrl = $parsedUrl['scheme'] . '://' . $parsedUrl['host'] . $parsedUrl['path'];
            $queryString = isset($parsedUrl['query']) ? $parsedUrl['query'] : '';
            
            // Check for preload/preconnect in request headers
            $isPreload = false;
            foreach ($request['headers'] as $header) {
                if (strtolower($header['name']) === 'purpose' && $header['value'] === 'prefetch') {
                    $isPreload = true;
                    break;
                }
            }
            
            // Determine resource type
            $resourceType = $this->getResourceType($url, $response);
            
            // Track URLs
            if (!isset($urlGroups[$url])) {
                $urlGroups[$url] = [];
            }
            $urlGroups[$url][] = [
                'method' => $method,
                'status' => $response['status'],
                'size' => $response['bodySize'] + $response['headersSize'],
                'mimeType' => $response['content']['mimeType'] ?? 'unknown',
                'resourceType' => $resourceType,
                'isPreload' => $isPreload
            ];
            
            // Track base URLs (without query string)
            if (!isset($baseUrlGroups[$baseUrl])) {
                $baseUrlGroups[$baseUrl] = [];
            }
            $baseUrlGroups[$baseUrl][] = [
                'fullUrl' => $url,
                'queryString' => $queryString,
                'status' => $response['status'],
                'size' => $response['bodySize'] + $response['headersSize']
            ];
            
            // Track preloads
            if ($isPreload) {
                $preloadUrls[$url] = true;
            }
            
            // Track fonts
            if ($resourceType === 'font') {
                if (!isset($fontUrls[$url])) {
                    $fontUrls[$url] = 0;
                }
                $fontUrls[$url]++;
            }
            
            // Track errors (4xx and 5xx status codes)
            if ($response['status'] >= 400) {
                $analysis['errorResponses'][] = [
                    'url' => $url,
                    'status' => $response['status'],
                    'statusText' => $response['statusText'] ?? '',
                    'resourceType' => $resourceType
                ];
            }
            
            // Count resources by type
            if (!isset($analysis['resourcesByType'][$resourceType])) {
                $analysis['resourcesByType'][$resourceType] = [
                    'count' => 0,
                    'totalSize' => 0
                ];
            }
            $analysis['resourcesByType'][$resourceType]['count']++;
            $analysis['resourcesByType'][$resourceType]['totalSize'] += $response['bodySize'] + $response['headersSize'];
        }
        
        // Find duplicate requests (exact same URL)
        foreach ($urlGroups as $url => $requests) {
            if (count($requests) > 1) {
                $totalSize = 0;
                foreach ($requests as $req) {
                    $totalSize += $req['size'];
                }
                $wastedSize = $totalSize - $requests[0]['size']; // All but first are wasted
                
                $analysis['duplicateRequests'][] = [
                    'url' => $url,
                    'count' => count($requests),
                    'resourceType' => $requests[0]['resourceType'],
                    'totalSize' => $totalSize,
                    'wastedBytes' => $wastedSize,
                    'requests' => $requests
                ];
                
                $analysis['wastedBytes'] += $wastedSize;
            }
            
            // Check if this URL was also preloaded
            if (isset($preloadUrls[$url]) && count($requests) > 1) {
                $analysis['preloadDuplicates'][] = [
                    'url' => $url,
                    'timesLoaded' => count($requests),
                    'resourceType' => $requests[0]['resourceType']
                ];
            }
        }
        
        // Find query string duplicates (same base URL, different query strings)
        foreach ($baseUrlGroups as $baseUrl => $requests) {
            if (count($requests) > 1) {
                $uniqueQueryStrings = [];
                foreach ($requests as $req) {
                    $uniqueQueryStrings[$req['queryString']] = true;
                }
                
                // Only report if there are different query strings
                if (count($uniqueQueryStrings) > 1) {
                    $analysis['queryStringDuplicates'][] = [
                        'baseUrl' => $baseUrl,
                        'variants' => array_map(function($req) {
                            return [
                                'url' => $req['fullUrl'],
                                'queryString' => $req['queryString'],
                                'size' => $req['size']
                            ];
                        }, $requests)
                    ];
                }
            }
        }
        
        // Find font duplicates
        foreach ($fontUrls as $url => $count) {
            if ($count > 1) {
                $analysis['fontDuplicates'][] = [
                    'url' => $url,
                    'timesLoaded' => $count
                ];
            }
        }
        
        return $analysis;
    }
    
    private function getResourceType($url, $response) {
        $mimeType = $response['content']['mimeType'] ?? '';
        $extension = pathinfo(parse_url($url, PHP_URL_PATH), PATHINFO_EXTENSION);
        
        // Check by MIME type first
        if (strpos($mimeType, 'javascript') !== false || strpos($mimeType, 'ecmascript') !== false) {
            return 'js';
        }
        if (strpos($mimeType, 'css') !== false) {
            return 'css';
        }
        if (strpos($mimeType, 'image/') !== false) {
            return 'image';
        }
        if (strpos($mimeType, 'font/') !== false || strpos($mimeType, 'application/font') !== false) {
            return 'font';
        }
        if (strpos($mimeType, 'text/html') !== false) {
            return 'html';
        }
        if (strpos($mimeType, 'application/json') !== false) {
            return 'json';
        }
        
        // Check by extension
        switch (strtolower($extension)) {
            case 'js':
            case 'mjs':
                return 'js';
            case 'css':
                return 'css';
            case 'jpg':
            case 'jpeg':
            case 'png':
            case 'gif':
            case 'svg':
            case 'webp':
            case 'ico':
                return 'image';
            case 'woff':
            case 'woff2':
            case 'ttf':
            case 'otf':
            case 'eot':
                return 'font';
            case 'html':
            case 'htm':
                return 'html';
            case 'json':
                return 'json';
            case 'xml':
                return 'xml';
            default:
                return 'other';
        }
    }
    
    public function formatReport($analysis) {
        $report = "\n=== HAR Analysis Report: {$analysis['fileName']} ===\n\n";
        
        // Summary
        $report .= "SUMMARY:\n";
        $report .= "- Total Requests: {$analysis['totalRequests']}\n";
        $report .= "- Duplicate Requests: " . count($analysis['duplicateRequests']) . "\n";
        $report .= "- Error Responses (4xx/5xx): " . count($analysis['errorResponses']) . "\n";
        $report .= "- Total Wasted Bytes: " . number_format($analysis['wastedBytes']) . " bytes (" . $this->formatBytes($analysis['wastedBytes']) . ")\n\n";
        
        // Resources by Type
        $report .= "RESOURCES BY TYPE:\n";
        foreach ($analysis['resourcesByType'] as $type => $data) {
            $report .= sprintf("- %s: %d requests, %s\n", 
                strtoupper($type), 
                $data['count'], 
                $this->formatBytes($data['totalSize'])
            );
        }
        $report .= "\n";
        
        // Duplicate Requests
        if (!empty($analysis['duplicateRequests'])) {
            $report .= "DUPLICATE REQUESTS:\n";
            foreach ($analysis['duplicateRequests'] as $dup) {
                $report .= sprintf("- %s\n  Type: %s, Loaded: %d times, Wasted: %s\n", 
                    $this->truncateUrl($dup['url']),
                    $dup['resourceType'],
                    $dup['count'],
                    $this->formatBytes($dup['wastedBytes'])
                );
            }
            $report .= "\n";
        }
        
        // Query String Duplicates
        if (!empty($analysis['queryStringDuplicates'])) {
            $report .= "QUERY STRING DUPLICATES (same file, different parameters):\n";
            foreach ($analysis['queryStringDuplicates'] as $dup) {
                $report .= "- Base URL: " . $this->truncateUrl($dup['baseUrl']) . "\n";
                foreach ($dup['variants'] as $variant) {
                    $report .= sprintf("  * %s (%s)\n", 
                        $variant['queryString'] ?: '[no query string]',
                        $this->formatBytes($variant['size'])
                    );
                }
            }
            $report .= "\n";
        }
        
        // Preload Duplicates
        if (!empty($analysis['preloadDuplicates'])) {
            $report .= "PRELOAD DUPLICATES:\n";
            foreach ($analysis['preloadDuplicates'] as $dup) {
                $report .= sprintf("- %s (Type: %s, Loaded: %d times)\n", 
                    $this->truncateUrl($dup['url']),
                    $dup['resourceType'],
                    $dup['timesLoaded']
                );
            }
            $report .= "\n";
        }
        
        // Font Duplicates
        if (!empty($analysis['fontDuplicates'])) {
            $report .= "FONT DUPLICATES:\n";
            foreach ($analysis['fontDuplicates'] as $font) {
                $report .= sprintf("- %s (Loaded: %d times)\n", 
                    $this->truncateUrl($font['url']),
                    $font['timesLoaded']
                );
            }
            $report .= "\n";
        }
        
        // Error Responses
        if (!empty($analysis['errorResponses'])) {
            $report .= "ERROR RESPONSES:\n";
            foreach ($analysis['errorResponses'] as $error) {
                $report .= sprintf("- %d %s: %s (Type: %s)\n", 
                    $error['status'],
                    $error['statusText'],
                    $this->truncateUrl($error['url']),
                    $error['resourceType']
                );
            }
            $report .= "\n";
        }
        
        return $report;
    }
    
    private function formatBytes($bytes) {
        if ($bytes < 1024) return $bytes . ' B';
        if ($bytes < 1048576) return round($bytes / 1024, 2) . ' KB';
        return round($bytes / 1048576, 2) . ' MB';
    }
    
    private function truncateUrl($url, $maxLength = 80) {
        if (strlen($url) <= $maxLength) return $url;
        
        $parsed = parse_url($url);
        $path = $parsed['path'] ?? '';
        $filename = basename($path);
        
        if (strlen($filename) < 40) {
            return '...' . substr($url, -($maxLength - 3));
        } else {
            return '...' . substr($filename, -($maxLength - 3));
        }
    }
}

// Main execution
$harFiles = [
    'trang-chu.har',
    'trang-san-pham.har',
    'trang-bai-viet.har',
    'trang-contact.har'
];

$inputDir = __DIR__ . '/inputs/';
$allReports = '';

foreach ($harFiles as $harFile) {
    $filePath = $inputDir . $harFile;
    
    if (!file_exists($filePath)) {
        echo "Warning: HAR file not found: $filePath\n";
        continue;
    }
    
    try {
        $analyzer = new HarAnalyzer($filePath);
        $analysis = $analyzer->analyze();
        $report = $analyzer->formatReport($analysis);
        
        echo $report;
        $allReports .= $report;
        
    } catch (Exception $e) {
        echo "Error analyzing $harFile: " . $e->getMessage() . "\n";
    }
}

// Save combined report
file_put_contents(__DIR__ . '/har-analysis-report.txt', $allReports);
echo "\nFull report saved to: " . __DIR__ . '/har-analysis-report.txt\n";